external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -225,6 +243,25 @@ private long doSeek(TopicPartition tp, OffsetAndMetadata committedOffset) {
         }
     }
 
+    /**
+     * Checks If {@link OffsetAndMetadata} was committed by an instance of {@link KafkaSpout} in this topology.
+     * This info is used to decide if {@link FirstPollOffsetStrategy} should be applied
+     *
+     * @param committedOffset {@link OffsetAndMetadata} info committed to Kafka
+     * @return true if this topology committed this {@link OffsetAndMetadata}, false otherwise
+     */
+    private boolean isOffsetCommittedByThisTopology(OffsetAndMetadata committedOffset) {
+        try {
+            final KafkaSpout.Info info = JSON_MAPPER.readValue(committedOffset.metadata(), KafkaSpout.Info.class);
+            return info.getTopologyId().equals(context.getStormId());
+        } catch (IOException e) {
+            LOG.trace("Failed to deserialize {}. Error likely occurred because the last commit " +
```

Comment:
I don't think anyone is going to see this log if it's left at trace level.

Reply:
If we put it at WARN level, it will write this message potentially thousands of times (equal to the number of records that get polled), until the first commit is done. I put it at TRACE because this is technically for debugging purposes, and I am printing the exception stack trace as well. I can put it at DEBUG, but I don't think it should be WARN. I can put the message at  DEBUG and stack trade at TRACE. 

On a side note, I don't think TRACE level is used enough, but I consider it is very useful to print more detail after the problem is identified.

Reply:
Good point, hadn't thought of this code being hit on every emit. The reason I think WARN level is appropriate is that this is indicating a potential error, and we want it to show up in the log if an exception is caught here. Putting it at TRACE level means no one will see it. It's not solely for debugging purposes, it's informational telling the user that the spout is defaulting to a certain behavior because a feature failed to work. TRACE and DEBUG are IMO for logs that happen "normally" and not special cases like this.

I'm wondering if we should cache the committed offset and metadata in a field between commits instead, i.e. add a partition -> lastCommitedOffsetAndMetadata map that we update in the commit method? It seems a little inefficient to run this code on every single emit, in addition to the issue with lots of logging.

Reply:
The log messages only get printed in the case of a commit made by a topology running an older version of Storm and until the first commit is done by this topology for each topic-partition. Ater that the first commit logs will no longer occur. I will look into if it makes sense caching this info.

Reply:
Sure, but we're still deserializing the metadata on every emit. I think we can very likely get away with deserializing once per commit, or even once per spout activation, since we don't need to serialize + deserialize if we only update the metadata when committing.